File: src-tauri/src/db/conflicts.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use crate::error::AppError;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Conflict {
    pub id: String,
    pub folder_id: String,
    pub file_path: String,
    pub local_version: Option<String>,
    pub remote_version: Option<String>,
    pub detected_at: String,
    pub resolved: bool,
    pub resolution: Option<String>,
}

pub fn list(conn: &Connection, resolved: Option<bool>) -> Result<Vec<Conflict>, AppError> {
    let query = match resolved {
        Some(true) => "SELECT id, folder_id, file_path, local_version, remote_version, detected_at, resolved, resolution FROM conflicts WHERE resolved = 1 ORDER BY detected_at DESC",
        Some(false) => "SELECT id, folder_id, file_path, local_version, remote_version, detected_at, resolved, resolution FROM conflicts WHERE resolved = 0 ORDER BY detected_at DESC",
        None => "SELECT id, folder_id, file_path, local_version, remote_version, detected_at, resolved, resolution FROM conflicts ORDER BY detected_at DESC",
    };

    let mut stmt = conn.prepare(query)?;
    let conflicts = stmt.query_map([], |row| {
        Ok(Conflict {
            id: row.get(0)?,
            folder_id: row.get(1)?,
            file_path: row.get(2)?,
            local_version: row.get(3)?,
            remote_version: row.get(4)?,
            detected_at: row.get(5)?,
            resolved: row.get(6)?,
            resolution: row.get(7)?,
        })
    })?.collect::<Result<Vec<_>, _>>()?;

    Ok(conflicts)
}

pub fn create(
    conn: &Connection,
    id: &str,
    folder_id: &str,
    file_path: &str,
    local_version: Option<&str>,
    remote_version: Option<&str>,
) -> Result<(), AppError> {
    conn.execute(
        "INSERT INTO conflicts (id, folder_id, file_path, local_version, remote_version) VALUES (?1, ?2, ?3, ?4, ?5)",
        params![id, folder_id, file_path, local_version, remote_version],
    )?;
    Ok(())
}

pub fn resolve(conn: &Connection, id: &str, resolution: &str) -> Result<(), AppError> {
    conn.execute(
        "UPDATE conflicts SET resolved = 1, resolution = ?1 WHERE id = ?2",
        params![resolution, id],
    )?;
    Ok(())
}

pub fn count_unresolved(conn: &Connection) -> Result<i64, AppError> {
    let count: i64 = conn.query_row(
        "SELECT COUNT(*) FROM conflicts WHERE resolved = 0",
        [],
        |row| row.get(0),
    )?;
    Ok(count)
}

File: server/src/routes/conflicts.ts
import { Hono } from "hono";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";
import type { Env } from "../env";
import { authMiddleware, type AuthContext } from "../middleware/auth";

const conflicts = new Hono<{ Bindings: Env; Variables: AuthContext }>();

// All conflict routes require authentication
conflicts.use("*", authMiddleware);

// GET /conflicts - List user's conflicts
conflicts.get("/", async (c) => {
  const userId = c.get("user").sub;
  const resolved = c.req.query("resolved");

  let query =
    "SELECT id, user_id, folder_id, file_path, local_version, remote_version, detected_at, resolved, resolution FROM conflicts WHERE user_id = ?1";

  const params: (string | number)[] = [userId];

  if (resolved === "true") {
    query += " AND resolved = 1";
  } else if (resolved === "false") {
    query += " AND resolved = 0";
  }

  query += " ORDER BY detected_at DESC";

  const rows = await c.env.DB.prepare(query).bind(...params).all();

  return c.json({ conflicts: rows.results });
});

// PATCH /conflicts/:id - Resolve a conflict
const resolveSchema = z.object({
  resolution: z.enum(["keep_local", "keep_remote", "keep_both"]),
});

conflicts.patch("/:id", zValidator("json", resolveSchema), async (c) => {
  const userId = c.get("user").sub;
  const conflictId = c.req.param("id");
  const { resolution } = c.req.valid("json");
  const db = c.env.DB;

  const existing = await db
    .prepare("SELECT id FROM conflicts WHERE id = ?1 AND user_id = ?2")
    .bind(conflictId, userId)
    .first();

  if (!existing) {
    return c.json({ error: "Conflict not found" }, 404);
  }

  await db
    .prepare("UPDATE conflicts SET resolved = 1, resolution = ?1 WHERE id = ?2")
    .bind(resolution, conflictId)
    .run();

  return c.json({ ok: true });
});

export default conflicts;
